source/common/network/io_socket_handle_impl.cc

```diff
@@ -131,5 +133,100 @@ IoSocketHandleImpl::sysCallResultToIoCallResult(const Api::SysCallSizeResult& re
+  for (uint64_t i = 0; i < num_slice; i++) {
+    if (slices[i].mem_ != nullptr && slices[i].len_ != 0) {
+      iov[num_slices_for_read].iov_base = slices[i].mem_;
+      iov[num_slices_for_read].iov_len = slices[i].len_;
+      ++num_slices_for_read;
+    }
+  }
+
+  sockaddr_storage peer_addr;
+  msghdr hdr;
+  hdr.msg_name = &peer_addr;
+  hdr.msg_namelen = sizeof(sockaddr_storage);
+  hdr.msg_iov = iov.begin();
+  hdr.msg_iovlen = num_slices_for_read;
+  hdr.msg_flags = 0;
+
+  struct cmsghdr* cmsg = reinterpret_cast<struct cmsghdr*>(cbuf);
+  cmsg->cmsg_len = cmsg_space;
+  hdr.msg_control = cmsg;
+  hdr.msg_controllen = cmsg_space;
+  auto& os_sys_calls = Api::OsSysCallsSingleton::get();
+  const Api::SysCallSizeResult result = os_sys_calls.recvmsg(fd_, &hdr, 0);
+  if (result.rc_ < 0) {
+    return sysCallResultToIoCallResult(result);
+  }
+  if (hdr.msg_flags & MSG_CTRUNC) {
+    ENVOY_LOG(error, "Incorrectly set control message length: {}", hdr.msg_controllen);
+    return sysCallResultToIoCallResult(result);
+  }
+
+  RELEASE_ASSERT(hdr.msg_namelen > 0,
+                 fmt::format("Unable to get remote address from recvmsg() for fd: {}", fd_));
+  try {
+    peer_address = Address::addressFromSockAddr(peer_addr, hdr.msg_namelen, v6only);
+  } catch (const EnvoyException& e) {
+    ENVOY_LOG(error, "Invalid remote address for fd: {}, error: {}", fd_, e.what());
+  }
+
+  // Get overflow, local and peer addresses from control message.
+  if (hdr.msg_controllen > 0) {
+    struct cmsghdr* cmsg;
+    for (cmsg = CMSG_FIRSTHDR(&hdr); cmsg != nullptr; cmsg = CMSG_NXTHDR(&hdr, cmsg)) {
+      if (cmsg->cmsg_type == IPV6_PKTINFO) {
+        struct in6_pktinfo* info = reinterpret_cast<in6_pktinfo*>(CMSG_DATA(cmsg));
+        sockaddr_in6 ipv6_addr;
+        memset(&ipv6_addr, 0, sizeof(sockaddr_in6));
+        ipv6_addr.sin6_family = AF_INET6;
+        ipv6_addr.sin6_addr = info->ipi6_addr;
+        ipv6_addr.sin6_port = htons(self_port);
+        local_address = Address::addressFromSockAddr(reinterpret_cast<sockaddr_storage&>(ipv6_addr),
+                                                     sizeof(sockaddr_in6), v6only);
+      }
+#if defined(IP_PKTINFO)
+      if (cmsg->cmsg_type == IP_PKTINFO) {
+        struct in_pktinfo* info = reinterpret_cast<in_pktinfo*>(CMSG_DATA(cmsg));
+#else
+      if (cmsgptr->cmsg_type == IP_RECVDSTADDR) {
+        struct in_addr* addr = reinterpret_cast<in_addr*>(CMSG_DATA(cmsg));
+#endif
+        sockaddr_in ipv4_addr;
+        memset(&ipv4_addr, 0, sizeof(sockaddr_in));
+        ipv4_addr.sin_family = AF_INET;
+        ipv4_addr.sin_addr =
+#if defined(IP_PKTINFO)
+            info->ipi_addr;
+#else
+            ipv4_addr.sin_addr = addr;
+#endif
+        ipv4_addr.sin_port = htons(self_port);
+        local_address = Address::addressFromSockAddr(reinterpret_cast<sockaddr_storage&>(ipv4_addr),
+                                                     sizeof(sockaddr_in), v6only);
+      }
```

Comment:
Can we put line 193-222 into a separate function, which takes `cmsg` and `self_port` as input and return a `InstanceConstSharedPtr` as output?

Similarly, put line 223-225 into a separate function which returns a `absl::optional<uint32_t>`.

Reply:
panic here also

Reply:
done